### PR TITLE
Pf/search lyrics only

### PR DESCRIPTION
**Pull request by patrickf3139**
Summary
- on song save, extract lyrics into own column
- update song fixtures to use GP header style (include : at the end) and manually extract lyrics
- improve song normalization to strip extra white spaces in the name, artist, and chord sheet fields

### CODE REVIEW

File: app/models/song.rb [hunk 1]
```diff
@@ -16,15 +16,35 @@ class Song < ActiveRecord::Base
...
+    self.chord_sheet.split("\n").each { |line| normalized_lines << line.rstrip }
+    self.chord_sheet = normalized_lines.join("\n")
+  end
+
+  def extract_lyrics
+    lyrics = self.chord_sheet.split("\n").find_all { |line| is_line_of_lyrics(line) }
+    self.lyrics = lyrics.join("\n")
+  end
+
+  CHORD_GIVEAWAY_STRINGS = Set.new(%w(# / [ 2 3 4 7 9))
```
Comment (leethomas) [inline app/models/song.rb]: not a big deal, but I prefer constants to be defined at the top of the file

File: app/models/song.rb [hunk 2]
```diff
@@ -16,15 +16,35 @@ class Song < ActiveRecord::Base
...
+  def is_line_of_lyrics(line)
+    # headers always end with :
+    return false if line[-1] == ":"
+    # chord lines contain these giveaway substrings
+    return false if line.split.any? {|word| CHORD_GIVEAWAY_STRINGS.include? word}
+    # chord lines have a high density of spaces or contain less than 3 words
+    return (line.count(" ") / line.length.to_f < 0.31) && (line.length > 6)
```
Comment (leethomas) [inline app/models/song.rb]: nice!

File: db/migrate/20161003203532_add_lyrics_column_to_songs.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# this migration will fail on databases that already contain songs
+class AddLyricsColumnToSongs < ActiveRecord::Migration
+  def change
+    add_column :songs, :lyrics, :text, null: false, after: :chord_sheet
```
Comment (leethomas) [inline db/migrate/20161003203532_add_lyrics_column_to_songs.rb]: If we're searching against this now then we'll want an index, gin type I think. Also should we remove the index from chord_sheet?

**Comment by natyconnor** [comment]
+1 for extracting the lyrics into it's own column. I think the next step will be to split the chords off into their own table, to allow transposing. Might have time this week to take a stab since we're taking a little break...

File: app/models/song.rb (hunk 1 above)
Comment (natyconnor) [inline app/models/song.rb]: If this was at work, I would say move the constant to the top of the private section then. But we're really obsessive about clean code and consistent formatting so up to you.

**Comment by leethomas** [comment]
@patrickf3139 @natyconnor there's also a class method called `private_constant`. So this can be done too:

```rb
class Song
  CHORD_GIVEAWAY_STRINGS = Set.new(%w(# / [ 2 3 4 7 9))
  private_constant :CHORD_GIVEAWAY_STRINGS
end
```

Not saying we need to change it, just fyi: https://ruby-doc.org/core-2.1.1/Module.html#method-i-private_constant

File: app/models/song.rb (hunk 1 above)
Comment (natyconnor) [inline app/models/song.rb]: Also, I'm guessing you're using the numbers from suspended or add chords for this, in which case technically the list should be [2, 4, 6, 7, 9, 11, 13, 14, ... ]

File: app/models/song.rb (hunk 2 above)
Comment (natyconnor) [inline app/models/song.rb]: Would this work for something like this:
```
C/E   F      C/E     F        Gm  F      C/E   Dm
One name holds weight a  -  bove them all
C/E  F       C/E   F     Gm   F     C/E  Dm
His   fame out - lasts the   earth He  formed
F/A  Bb      F/A   Bb  C  Bb   F/A  Dm
His   praise resounds beyond the  stars
```
This is a pretty extreme case but just want to make sure the density isn't too loose. 

Also, I'm pretty sure there's a number of songs with 3 or less word length lines. No other name again has a nice edge case:

```
     F
Jesus
```

**Comment by patrickf3139** [comment]
Haha you guys are destroying my PR while I'm at Joyland workday. Will address them after :)

**Comment by patrickf3139** [comment]
Also, @natyconnor , feel free to review https://github.com/GracepointMinistries/GraceTunes/pull/12 even though it's already been merged. I'm not great at FE so I didn't really fully review it.

**Comment by patrickf3139** [comment]
@natyconnor So let me understand what you suggested about having a Chords table. You think we should create a Chords table where there is one column for every possible transposition of every chord sheet so we can just read from that table when users request a transposal?
Regardless of what you mean, the real difficult part of transposing that we'll need to do is how to take a chord sheet in the original key and transpose all the chords while retaining the format of the chord sheet. If you're free, feel free to tackle that problem.

File: app/models/song.rb (hunk 1 above)
Comment (patrickf3139) [inline app/models/song.rb]: Sure I'll move it up. Yeah these numbers for for suspended chords. Thanks for the correction. Neither Lee nor I know much music theory so your understanding of both will be very valuable to us.

**Comment by natyconnor** [comment]
@patrickf3139 I'll put something in asana, but yeah, while I was thinking about it, how the chords relate to the song is the hard part. My initial thought was a Chord would belong_to a song, and also have a position of some kind in the song. Perhaps we need to have some kind of measure of beats or something in a song? Anyway, we can move this to a real issue, but I think that would make the design a lot more flexible, since right now the chords are just text in the song, and not an actual entity of some kind, so they're more or less baked in.

**Comment by patrickf3139** [comment]
Ok we can discuss the design on Asana or somewhere or in person but I think it's fine to just have chords baked into the song sheet. If we use these two [regexes](https://gist.github.com/andrewstucki/106c9704be9233e197350ceabec6a32c#file-parser-rb-L19) Andrew Stucki wrong to tokenize the chords, we can easily use a Map to change each chord while rewriting the chord sheet. Change in space might be difficult if the chords are placed close together, but that is an edge case we can deal with in the future. Having a row in the database for each chord in a chord sheet sounds like overkill to me.

Also, are you okay if we ignore that edge case with lyrics extraction and just merge? I will improve the algorithm when we have more actual data (which Steven Chang is working on) to test and iterate on.

**Comment by natyconnor** [comment]
Wow, that parser pretty much does everything we want right? It feels kinda hackish to just manually parse text for everything but if the work's already done then we might as well use it. Has transposing and everything already.

**Comment by patrickf3139** [comment]
Thanks gonna merge. Um Andrew told me it's completely untested, which is why I didn't just use that regex to do the lyrics extraction. But I do plan on using a revised version of it in the future.
